### PR TITLE
Property-based test: lwaftr run does not crash with get

### PR DESCRIPTION
**Pull request by kbara**
This adds a first property based test. The property is that the lwaftr
does not crash with a 'snabb config get' command run repeatedly.

This is a first step towards resolving #631 .

Use: ```SNABB_PCI0=82:00.1 ./program/lwaftr/tests/propbased/selftest.sh```

Note that this is currently reporting crashes, for instance after 14 or 23 iterations.

### CODE REVIEW

**Comment by takikawa** [comment]
I've been poking around to see if I can figure out why it crashes, but had limited success so far. Adding some prints produces this error message:

```
core/main.lua:26: short read
stack traceback:
	core/main.lua:149: in function <core/main.lua:147>
	[C]: in function 'error'
	core/main.lua:26: in function 'assert'
	program/config/common.lua:121: in function 'read_length'
	program/config/common.lua:144: in function 'recv_message'
	program/config/common.lua:153: in function 'call_leader'
	program/config/get/get.lua:8: in function 'run'
	program/config/config.lua:23: in function 'run'
	core/main.lua:67: in function <core/main.lua:43>
	[C]: in function 'xpcall'
	core/main.lua:211: in main chunk
	[C]: at 0x004535b0
	[C]: in function 'pcall'
	core/startup.lua:3: in main chunk
	[C]: in function 'require'
	[string "require "core.startup""]:1: in main chunk
```

So it looks like the config program is getting EOF at the point where it tries to read the message length part of the protocol right after sending a get-config message (the describe message seems to get through).

BTW, while reading through some of the RPC code I think I found a minor bug but not 100% sure. Here's a patch for it included below:

```
From fbff900411bc6e52ad4ca83ee6037f78100e7b60 Mon Sep 17 00:00:00 2001
From: Asumu Takikawa <asumu@igalia.com>
Date: Thu, 8 Dec 2016 00:07:43 -0800
Subject: [PATCH] Fix accidental hardcoding of the schema name

---
 src/lib/yang/rpc.lua | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/lib/yang/rpc.lua b/src/lib/yang/rpc.lua
index 0d2fde5..6c5395b 100644
--- a/src/lib/yang/rpc.lua
+++ b/src/lib/yang/rpc.lua
@@ -15,7 +15,7 @@ function prepare_callee(schema_name)
 end
 
 function prepare_caller(schema_name)
-   local schema = schema.load_schema_by_name('snabb-config-leader-v1')
+   local schema = schema.load_schema_by_name(schema_name)
    return {
       print_input = data.rpc_input_printer_from_schema(schema),
       parse_output = data.rpc_output_parser_from_schema(schema)
-- 
2.1.4

```

**Comment by wingo** [comment]
Hi, "short read" means the leader hung up on the call, which it will do if there is any error.  We have to look on the leader side for the details.  The patch isn't right -- we always talk to the leader with the snabb-config-leader-v1 rpcs, they embed data for other schemas (see that yang model).

File: src/program/lwaftr/tests/propbased/selftest.sh [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env bash
+
+set -e
+
+SKIPPED_CODE=43
+
+if [ -z $SNABB_PCI0 ]; then exit SKIPPED_CODE; fi
```
Comment (teknico) [inline src/program/lwaftr/tests/propbased/selftest.sh]: Dollar sign missing: `SKIPPED_CODE` -> `$SKIPPED_CODE`

Comment (kbara) [reply src/program/lwaftr/tests/propbased/selftest.sh]: Good catch, that line used to have two bugs.

File: src/program/lwaftr/tests/propbased/prop_nocrash.lua [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+   local pid = S.fork()
+   if pid == 0 then
+      local cmdline = {"snabb", "lwaftr", "run", "-D", "2", "--conf",
+          "program/lwaftr/tests/data/icmp_on_fail.conf", "--reconfigurable", 
+          "--on-a-stick", pci_addr}
+      -- FIXME: preserve the environment
+      S.execve(("/proc/%d/exe"):format(S.getpid()), cmdline, {})
+   else
+      run_pid = pid
+      S.sleep(0.1)
```
Comment (teknico) [inline src/program/lwaftr/tests/propbased/prop_nocrash.lua]: It looks like this sleep is too short. When running `selftest.sh` I always get (shortened):

```
Could not connect to config leader socket on Snabb instance '4675'.
Launching snabb run failed, or we've crashed it!
Git revision c6a2f48942ef74638e0bf49627ad62ecc2c073b7
The property was falsified.
The command was:	./snabb config get 4675 /
Rerun as: snabb quickcheck --seed=458007436 --iterations=14 \
  program.lwaftr.tests.propbased.prop_nocrash 83:00.1
```

Then, if I run `snabb quickcheck` directly, then it works, but not always.

However, if I lengthen the sleep to 1s, then `snabb quickcheck` always works. `selftest.sh` still never does, though.

**Comment by takikawa** [comment]
After @wingo figured out that there was some GC trouble going on (adding `collectgarbage()` makes it crash quickly), I kept looking and it seems like the problem is that `peer.buf` gets freed by the GC too soon because `peer` itself is freed after the write.

Here are two patches that at least prevent the crashing. Not proposing this as a real fix though, since it's a kludge.

Patch 1:

```
From 53e93b6036e7942eb7bcc1b8b4cdce8c564bbb96 Mon Sep 17 00:00:00 2001
From: Asumu Takikawa <asumu@igalia.com>
Date: Fri, 9 Dec 2016 00:01:50 +0000
Subject: [PATCH 1/2] Hang onto peer.buf to avoid GC

Not a real solution, but prevents crashing
---
 src/apps/config/leader.lua | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/apps/config/leader.lua b/src/apps/config/leader.lua
index 53dcde7..23f03e3 100644
--- a/src/apps/config/leader.lua
+++ b/src/apps/config/leader.lua
@@ -532,6 +532,7 @@ function Leader:handle (payload)
 end
 
 local dummy_unix_sockaddr = S.t.sockaddr_un()
+local bufs_table = {}
 
 function Leader:handle_calls_from_peers()
    local peers = self.peers
@@ -608,6 +609,7 @@ function Leader:handle_calls_from_peers()
             reply = #reply..'\n'..reply
             peer.state = 'reply'
             peer.buf = ffi.new('uint8_t[?]', #reply, reply)
+	    table.insert(bufs_table, peer.buf)
             peer.pos = 0
             peer.len = #reply
          else
-- 
2.5.2


```

Patch 2:

```
From 61bd0817397d5d9a5a7aceb40f7ad6d944d8b0d8 Mon Sep 17 00:00:00 2001
From: Asumu Takikawa <asumu@igalia.com>
Date: Fri, 9 Dec 2016 00:04:06 +0000
Subject: [PATCH 2/2] Increase test duration for prop_nocrash

---
 src/program/lwaftr/tests/propbased/prop_nocrash.lua | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/program/lwaftr/tests/propbased/prop_nocrash.lua b/src/program/lwaftr/tests/propbased/prop_nocrash.lua
index 136a3f9..62afe9d 100644
--- a/src/program/lwaftr/tests/propbased/prop_nocrash.lua
+++ b/src/program/lwaftr/tests/propbased/prop_nocrash.lua
@@ -31,7 +31,7 @@ function handle_prop_args(prop_args)
 
    local pid = S.fork()
    if pid == 0 then
-      local cmdline = {"snabb", "lwaftr", "run", "-D", "2", "--conf",
+      local cmdline = {"snabb", "lwaftr", "run", "-D", "8", "--conf",
           "program/lwaftr/tests/data/icmp_on_fail.conf", "--reconfigurable", 
           "--on-a-stick", pci_addr}
       -- FIXME: preserve the environment
-- 
2.5.2


```

**Comment by takikawa** [comment]
The following patch is potentially a better fix, though I'm not entirely sure why it works. (if anyone who has a better grasp of LuaJIT's FFI semantics can take a look that'd be great) 

```
From 2023ccc8baaff8e7dc127ea7a2d32611dbb596dd Mon Sep 17 00:00:00 2001
From: Asumu Takikawa <asumu@igalia.com>
Date: Fri, 9 Dec 2016 05:16:23 +0000
Subject: [PATCH] Cast reply string instead of using ffi.new

---
 src/apps/config/leader.lua                          | 2 +-
 src/program/lwaftr/tests/propbased/prop_nocrash.lua | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/apps/config/leader.lua b/src/apps/config/leader.lua
index 53dcde7..ec9179b 100644
--- a/src/apps/config/leader.lua
+++ b/src/apps/config/leader.lua
@@ -607,7 +607,7 @@ function Leader:handle_calls_from_peers()
             assert(type(reply) == 'string')
             reply = #reply..'\n'..reply
             peer.state = 'reply'
-            peer.buf = ffi.new('uint8_t[?]', #reply, reply)
+            peer.buf = ffi.cast("uint8_t*", reply)
             peer.pos = 0
             peer.len = #reply
          else
diff --git a/src/program/lwaftr/tests/propbased/prop_nocrash.lua b/src/program/lwaftr/tests/propbased/prop_nocrash.lua
index 136a3f9..6743749 100644
--- a/src/program/lwaftr/tests/propbased/prop_nocrash.lua
+++ b/src/program/lwaftr/tests/propbased/prop_nocrash.lua
@@ -31,7 +31,7 @@ function handle_prop_args(prop_args)
 
    local pid = S.fork()
    if pid == 0 then
-      local cmdline = {"snabb", "lwaftr", "run", "-D", "2", "--conf",
+      local cmdline = {"snabb", "lwaftr", "run", "-D", "10", "--conf",
           "program/lwaftr/tests/data/icmp_on_fail.conf", "--reconfigurable", 
           "--on-a-stick", pci_addr}
       -- FIXME: preserve the environment
-- 
2.5.2


```

**Comment by wingo** [comment]
Aaaaah, excellent work @takikawa.  So it turns out it's not an early free but rather corruption.  Quoth http://luajit.org/ext_ffi_semantics.html in "Initializers":

> Byte arrays may also be initialized with a Lua string. This copies the whole string plus a terminating zero-byte. The copy stops early only if the array has a known, fixed size.

I was thinking that `ffi.new('uint8_t[?]', #str, str)` would make an array with "known, fixed size".  But I think that characteristic applies to the type, and is `uint8_t[?]` fixed, even when given its length?  I thought so.  I was wrong; you can't do this:

```lua
local buf = ffi.typeof('uint8_t[?]', #str)(str)
```

But you can do this:

```lua
local buf = ffi.typeof('uint8_t[$]', #str)(str)
```

Using the `$` makes an array type with fixed size.  The `[?]` makes a variable-length array (VLA).

So this bug was that LuaJIT was writing a zero byte beyond the array, and that was corrupting the GC metadata (probably the header for the next object).  Putting all buffers in an array mitigated the bug by never requiring that the data be collected.  The real fix tho is to stop writing beyond the end of data :)  I will prepare a PR.

**Comment by dpino** [comment]
Superseded by #703

**Pull request closed without merging**